File: csharp/exercise/exer02a02-producer-consumer.cs
/*
 * THE PRODUCER-CONSUMER PROBLEM
 *
 * SOLUTION TYPE A: USING BLOCKING QUEUES
 *      Version A02: 2 slow producers, 1 fast consumer
 */
using System;
using System.Collections.Concurrent;
using System.Threading;



class Exer02A02 : IRunnable
{
    public void run()
    {
        var queue = new BlockingCollection<int>(1);

        new Thread(() => producer(queue)).Start();
        new Thread(() => producer(queue)).Start();

        new Thread(() => consumer(queue)).Start();
    }


    private void producer(BlockingCollection<int> queue)
    {
        int i = 1;

        for (; ; ++i)
        {
            queue.Add(i);
            Thread.Sleep(1000);
        }
    }


    private void consumer(BlockingCollection<int> queue)
    {
        for (; ; )
        {
            int data = queue.Take();
            Console.WriteLine("Consumer " + data);
        }
    }
}

File: python/exer02a02_producer_consumer.py
'''
THE PRODUCER-CONSUMER PROBLEM

SOLUTION TYPE A: USING BLOCKING QUEUES
    Version A02: 2 slow producers, 1 fast consumer
'''

import time
from queue import Queue
import threading



def producer(q: Queue):
    i = 1

    while True:
        q.put(i)
        time.sleep(1)
        i += 1



def consumer(q: Queue):
    while True:
        data = q.get()
        print('Consumer', data)



blkq = Queue()

threading.Thread(target=producer, args=(blkq,)).start()
threading.Thread(target=producer, args=(blkq,)).start()

threading.Thread(target=consumer, args=(blkq,)).start()
